tutorials/advanced-topics/basic-segment-tree.md

```diff
@@ -0,0 +1,179 @@
+---
+
+title: 'Segment Tree'
+
+description: 'Author: @Dhruvil-Lakhtaria'
+
+hide_table_of_contents: true
+
+---
+
+<TutorialAuthors  names="@Dhruvil-Lakhtaria"/>  
+
+### Introduction
+  Segment Tree is a type of datastructure used in very specific kind of problems. Problems that deal with 2 types of operations namely point `update()` and `query()`(Some kind of queries like Max, Min, GCD, etc.) for a given array.
+
+### Structure
+It is basically a binary tree with 3 types of nodes - 
+
+ 1. `Root Node` - denotes the whole array.
+ 2. `Leaf Node` - denotes a single element of the array.
+ 3. `Internal Node` - denotes a sub-interval of the array.
+ 
+### Representation
+We can implement the above using an array representation.
+Each Index denotes the answer to that interval with $$(2*i+1) th$$ index as its left child and $$(2*i+2)th$$ index as its right child.
+
+![Segment-Tree](https://miro.medium.com/max/1100/1*9V1l8_weUfmT4jHg21EE5w.png)
+
+
+
+### Segment Tree Implementation for Range Min Query.
+
+Lets look at [307. Range Sum Query - Mutable](https://leetcode.com/problems/range-sum-query-mutable/) .
+> You are given an array with n values. You have to answer two type of operations on it.
+>
+> 1. Get sum of values in range $$[L-R]$$.
+> 2. Update ith element to new value. 
+
+### Approach Using Segment Tree.
+The question asks us to do 2 operations with the query operation having the logic to get sum of that operation. Seeing this only we can decide to use a segment tree data structure to solve the problem.
+
+### Building Segment tree for this question. 
+We start from the root denoting the entire array. Now we need information of $(0 - mid)$ and $(mid + 1 - n)$ for building current root. so we recursively call them. We know left child id equivalent to $2*i + 1$ array index and right child is $2*i + 2$.
+This recursive call stop when we reach a leaf denoting the single element. So we just update the array value to the elements value and backtrack.
+
+### Point updating the value.
+The logic for point update is pretty much the same involving recursive calls and backtracking. We reach till a leaf node where we need to update the value and while backtracking all the nodes that come into path needs to be checked for updation.
+
+For example, Analyse the image below.
+![Point Update](https://cdn.codespeedy.com/wp-content/uploads/2019/12/Node-or-Point-updates.png)
+
+Here we are update 2nd index and adding 3 to it.So all the nodes on the path back to root from it must also be updated.
+
+<Tabs>
+<TabItem value="CPP" label="CPP">
+<SolutionAuthor name="@DhruvilLakhtaria"/>
+
+```CPP
```

Comment:
format your code well following [this guideline](https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr)